tutorials/generative/2d_super_resolution/2d_stable_diffusion_v2_super_resolution-lightning.py

```diff
@@ -0,0 +1,552 @@
+                          
+# initialise the LightningModule
+ae_net = AutoEnconder()
+
+# set up checkpoints
+
+checkpoint_callback = ModelCheckpoint(dirpath=root_dir, filename="best_metric_model")
+
+                         
+# initialise Lightning's trainer.
+trainer = pl.Trainer(devices=1,
+                     max_epochs=n_epochs,
+                     check_val_every_n_epoch=val_interval,
+                     num_sanity_val_steps=0,
+                     callbacks=checkpoint_callback,
+                     default_root_dir=root_dir)
+
+# train
+trainer.fit(ae_net)
+
+
+# %% [markdown]
+# ## Rescaling factor
+#
+# As mentioned in Rombach et al. [1] Section 4.3.2 and D.1, the signal-to-noise ratio (induced by the scale of the latent space) became crucial in image-to-image translation models (such as the ones used for super-resolution). For this reason, we will compute the component-wise standard deviation to be used as scaling factor.
+
+# %%
+def get_scaler_factor():
+    ae_net.eval()
+    device = torch.device("cuda:0")
+    ae_net.to(device)
+
+    train_loader = ae_net.train_dataloader()
+    check_data = first(train_loader)
+    z = ae_net.autoencoderkl.encode_stage_2_inputs(check_data["image"].to(ae_net.device))
+    print(f"Scaling factor set to {1/torch.std(z)}")
+    scale_factor = 1 / torch.std(z)
+    return scale_factor
+
+scale_factor = get_scaler_factor()
+
+
+# %% [markdown]
+# ## Define the LightningModule for DiffusionModelUnet (transforms, network, loaders, etc)
+# The LightningModule contains a refactoring of your training code. The following module is a reformatiing of the code in 2d_stable_diffusion_v2_super_resolution.
+
+# %%
+class DiffusionUNET(pl.LightningModule):
+    def __init__(self):
+        super().__init__()
+        self.data_dir = root_dir
+        self.unet = DiffusionModelUNet(
+                                spatial_dims=2,
+                                in_channels=4,
+                                out_channels=3,
+                                num_res_blocks=2,
+                                num_channels=(256, 256, 512, 1024),
+                                attention_levels=(False, False, True, True),
+                                num_head_channels=(0, 0, 64, 64),
+                            )
+        self.max_noise_level = 350
+        self.scheduler = DDPMScheduler(num_train_timesteps=1000, 
+                          beta_schedule="linear",
+                          beta_start=0.0015,
+                          beta_end=0.0195)
+        self.z = ae_net.autoencoderkl.eval()
+
+
+    def forward(self, x, timesteps, low_res_timesteps):
+        return self.unet(x=x, 
+                         timesteps=timesteps,
+                         class_labels=low_res_timesteps)
+    
+           
+    def prepare_data(self):
+        self.train_ds, self.val_ds = get_datasets()
+        
+    def train_dataloader(self):
+        return DataLoader(self.train_ds, batch_size=16, shuffle=True,
```

Comment:
i get an error because `DataLoader` is never imported; only `ThreadDataLoader` is

Reply:
fixed that too!